src/topo.rs

```diff
@@ -0,0 +1,565 @@
+                target_stack.push(some_t);
+
+                t = if !target_que.is_empty() {
+                    // TODO:
+                    // use `target_que.pop_first()`
+                    let tmp = target_que.iter().next().copied().unwrap();
+                    target_que.remove(&tmp);
+
+                    Some(tmp.1)
+                } else {
+                    // paper says "Then we are done"
+                    None
+                };
+
+                // We don't want to return early and since we don't care about
+                // this number if `t` is None
+                t_outdegree = t.map_or(0, |id| {
+                    self.nodes.get(&id).map(|n| n.out_degree()).unwrap_or(0)
+                });
+            }
+        }
+
+        // Relabel the nodes...
+
+        if s.map_or(true, |s| s == to) {
+            s = self.previous(to);
+        }
+        if t.map_or(true, |t| t == from) {
+            t = self.next(from);
+        }
+
+        for s_first in source_stack.drain(..).rev() {
+            self.insert_after(s_first, s)?;
+            s = Some(s_first);
+        }
+
+        for t_last in target_stack.drain(..).rev() {
+            self.insert_before(t_last, t)?;
+            t = Some(t_last);
+        }
+
+        let s = s.ok_or(TopoError::NotFound)?;
+        let t = t.ok_or(TopoError::NotFound)?;
+        self.add_edge_to_graph(s, t)
+    }
+
+    fn add_edge_to_graph(&mut self, from: Id, to: Id) -> TopoResult<()> {
+        self.nodes
+            .get_mut(&from)
+            .ok_or(TopoError::NotFound)?
+            .add_outgoing_edge(to);
+        self.nodes
+            .get_mut(&to)
+            .ok_or(TopoError::NotFound)?
+            .add_incoming_edge(from);
+        Ok(())
+    }
+
+    fn is_before(&self, from: Id, to: Id) -> TopoResult<bool> {
+        let s = self.nodes.get_index_of(&from).ok_or(TopoError::NotFound)?;
+        let t = self.nodes.get_index_of(&to).ok_or(TopoError::NotFound)?;
+        Ok(s < t)
+    }
+
+    fn previous(&self, node: Id) -> Option<Id> {
+        self.nodes
+            .get_index_of(&node)
+            .and_then(|idx| idx.checked_sub(1))
+            .and_then(|idx| self.nodes.get_index(idx).map(|(k, _)| *k))
+    }
+
+    fn next(&self, node: Id) -> Option<Id> {
+        self.nodes
+            .get_index_of(&node)
+            .and_then(|idx| idx.checked_add(1))
+            .and_then(|idx| self.nodes.get_index(idx).map(|(k, _)| *k))
+    }
+
+    /// All edges to the given node.
+    ///
+    /// This represents all the nodes that can be reached by following incoming edges.
+    fn nodes_with_edges_incoming(&self, node: Id) -> Vec<(usize, Id)> {
+        self.nodes
+            .get(&node)
+            .map(|node| {
+                node.incoming
+                    .iter()
+                    .map(|id| Some((self.nodes.get_index_of(id)?, *id)))
+                    .flatten()
+            })
+            .into_iter()
+            .flatten()
```

Comment:
```suggestion
            .into_iter()
            .flat_map(|node| {
                node.incoming
                    .iter()
                    .filter_map(|id| Some((self.nodes.get_index_of(id)?, *id)))
            })
```

Also maybe return `impl Iterator` instead of collecting to `Vec`? Or does that not work?

Reply:
I think this `Some((self.nodes.get_index_of(id)?, *id))` is the problem of making it `impl Iterator` the (re?)borrow of `self` and moving it into the closure.